catalog: add Index.InvertedColumnKeyType

RaduBerinde opened the pull request:
In an inverted index, the encoded key is treated as a Bytes datum.
This is currently hardcoded in the opt catalog.

This commit moves this into the index descriptor.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75427)
<!-- Reviewable:end -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75427#-:-Mu8xtiHAx_xoh3eTIVM:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/catalog/table_elements.go, line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75427#-Mu8xb-A37oIcvtjupyO:-Mu8xb-A37oIcvtjupyP:b-4bbn1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/948765b7e220a4a168abd5b9185171281bfbc551/pkg/sql/catalog/table_elements.go#L157)):*
> ```Go
> 	// InvertedColumnID returns the ColumnID of the inverted column of the
> 	// inverted index.
> 	InvertedColumnID() descpb.ColumnID
> ```

Comment about the panic if it's not an inverted index
___
*[pkg/sql/catalog/table_elements.go, line 161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75427#-Mu8xe7q1EC0Jyrt8CFQ:-Mu8xe7q1EC0Jyrt8CFR:b-4bbn1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/948765b7e220a4a168abd5b9185171281bfbc551/pkg/sql/catalog/table_elements.go#L161)):*
> ```Go
> 	// InvertedColumnName returns the name of the inverted column of the inverted
> 	// index.
> 	InvertedColumnName() string
> ```

Comment about the panic if it's not an inverted index
___
*[pkg/sql/catalog/table_elements.go, line 165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75427#-Mu8xefo4vIdeWP367eh:-Mu8xefo4vIdeWP367ei:b-4bbn1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/948765b7e220a4a168abd5b9185171281bfbc551/pkg/sql/catalog/table_elements.go#L165)):*
> ```Go
> 	// InvertedColumnKeyType returns the type of the data element that is encoded
> 	// as the inverted index key. This is currently always Bytes.
> 	InvertedColumnKeyType() *types.T
> ```

Comment about the panic if it's not an inverted index


<!-- Sent from Reviewable.io -->

ajwerner commented:
rebase to fix the flake

RaduBerinde committed:
catalog: add Index.InvertedColumnKeyType

In an inverted index, the encoded key is treated as a Bytes datum.
This is currently hardcoded in the opt catalog.

This commit moves this into the index descriptor.

Release note: None

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75427#-:-Mu9BY5q2OsgsaUP5W4F:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @rytaft)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75427#-:-Mu9BY5q2OsgsaUP5W4F:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @rytaft)



<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75427#-:-MuBDbOj4mLUGPByK5s2:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTRs!

bors r+
